File: Client/MsgGen.java
import static java.lang.Math.round;

class MsgGen {
    // Door message codes
    private final static int LEFT_DOOR_IN = 1;
    private final static int RIGHT_DOOR_IN = 2;
    private final static int LEFT_DOOR_OUT = 3;
    private final static int RIGHT_DOOR_OUT = 4;
    private final static int LEFT_DOOR_STOP = 5;
    private final static int RIGHT_DOOR_STOP = 6;
    private final static int BOTH_DOORS_IN = 7;
    private final static int BOTH_DOORS_OUT = 8;
    private final static int BOTH_DOORS_STOP = 9;

    // Slope message codes
    private final static int SLOPE_STOP = 0;
    private final static int SLOPE_UP = 1;
    private final static int SLOPE_DOWN = 2;

    // Body codes
    private final static String DOOR_BODY = "d";
    private final static String SLOPE_BODY = "s";
    private final static String WHEELS_BODY = "w";
    private final static String LED_BODY = "l";
    private final static String GIMBAL_BODY = "g";

    // "API" codes
    // Doors
    public static final int D_LEFT = 0;
    public static final int D_RIGHT = 1;
    public static final int D_BOTH = 2;

    public static final int D_IN = 3;
    public static final int D_OUT = 4;
    public static final int D_STOP = 5;

    // Slope
    public static final int S_UP = 0;
    public static final int S_DOWN = 1;
    public static final int S_STOP = 2;

    // Gimbal
    public static final int G_STOP = 0;
    public static final int G_LEFT = 2;
    public static final int G_RIGHT = 3;
    public static final int G_UP = 4;
    public static final int G_DOWN = 5;


    private static String createMsg(String body, int code, int speed) {
        return "#" + body + "@" + code + "@" + speed + "!";
    }

    // "API" Functions
    public static String door(int side, int direction) {
        int instruction = -1;

        switch (side) {
            case D_LEFT:
                switch (direction) {
                    case D_IN:
                        instruction = LEFT_DOOR_IN;
                        break;
                    case D_OUT:
                        instruction = LEFT_DOOR_OUT;
                        break;
                    case D_STOP:
                        instruction = LEFT_DOOR_STOP;
                        break;
                }
                break;
            case D_RIGHT:
                switch (direction) {
                    case D_IN:
                        instruction = RIGHT_DOOR_IN;
                        break;
                    case D_OUT:
                        instruction = RIGHT_DOOR_OUT;
                        break;
                    case D_STOP:
                        instruction = RIGHT_DOOR_STOP;
                        break;
                }
                break;
            case D_BOTH:
                switch (direction) {
                    case D_IN:
                        instruction = BOTH_DOORS_IN;
                        break;
                    case D_OUT:
                        instruction = BOTH_DOORS_OUT;
                        break;
                    case D_STOP:
                        instruction = BOTH_DOORS_STOP;
                        break;
                }
                break;
        }
        return createMsg(DOOR_BODY, instruction, 0);
    }

    public static String slope(int direction) {
        int instruction = -1;

        switch(direction) {
            case S_UP:
                instruction = SLOPE_UP;
                break;
            case S_DOWN:
                instruction = SLOPE_DOWN;
                break;
            case S_STOP:
                instruction = SLOPE_STOP;
                break;
        }
        return createMsg(SLOPE_BODY, instruction, 0);
    }

    public static String wheels(double verticalSpeed, double horizontalSpeed) {
        return createMsg(WHEELS_BODY, 50 + (int) round(50.0*verticalSpeed), 50 + (int) round(50.0*horizontalSpeed));
    }

    public static String led() {
        return createMsg(LED_BODY, 0, 0);
    }

    public static String gimbal(int state) {
        return createMsg(GIMBAL_BODY, state, 0);
    }
}

File: GUI/app/src/main/java/engineering/design/rodofsteel/MsgGen.kt
package engineering.design.rodofsteel

/**
 * Created by s166315 on 15-10-2017.
 */
import java.lang.Math.round

internal object MsgGen {
    // Door message codes
    private val LEFT_DOOR_IN = 1
    private val RIGHT_DOOR_IN = 2
    private val LEFT_DOOR_OUT = 3
    private val RIGHT_DOOR_OUT = 4
    private val LEFT_DOOR_STOP = 5
    private val RIGHT_DOOR_STOP = 6
    private val BOTH_DOORS_IN = 7
    private val BOTH_DOORS_OUT = 8
    private val BOTH_DOORS_STOP = 9

    // Slope message codes
    private val SLOPE_STOP = 0
    private val SLOPE_UP = 1
    private val SLOPE_DOWN = 2

    // Body codes
    private val DOOR_BODY = "d"
    private val SLOPE_BODY = "s"
    private val WHEELS_BODY = "w"
    private val LED_BODY = "l"
    private val GIMBAL_BODY = "g"

    // "API" codes
    // Doors
    val D_LEFT = 0
    val D_RIGHT = 1
    val D_BOTH = 2

    val D_IN = 3
    val D_OUT = 4
    val D_STOP = 5

    // Slope
    val S_UP = 0
    val S_DOWN = 1
    val S_STOP = 2

    // Gimbal
    val G_STOP = 0
    val G_LEFT = 2
    val G_RIGHT = 3
    val G_UP = 4
    val G_DOWN = 5


    private fun createMsg(body: String, code: Int, speed: Int): String {
        return "#$body@$code@$speed!"
    }

    // "API" Functions
    fun door(side: Int, direction: Int): String {
        var instruction = -1

        when (side) {
            D_LEFT -> when (direction) {
                D_IN -> instruction = LEFT_DOOR_IN
                D_OUT -> instruction = LEFT_DOOR_OUT
                D_STOP -> instruction = LEFT_DOOR_STOP
            }
            D_RIGHT -> when (direction) {
                D_IN -> instruction = RIGHT_DOOR_IN
                D_OUT -> instruction = RIGHT_DOOR_OUT
                D_STOP -> instruction = RIGHT_DOOR_STOP
            }
            D_BOTH -> when (direction) {
                D_IN -> instruction = BOTH_DOORS_IN
                D_OUT -> instruction = BOTH_DOORS_OUT
                D_STOP -> instruction = BOTH_DOORS_STOP
            }
        }
        return createMsg(DOOR_BODY, instruction, 0)
    }

    fun slope(direction: Int): String {
        var instruction = -1

        when (direction) {
            S_UP -> instruction = SLOPE_UP
            S_DOWN -> instruction = SLOPE_DOWN
            S_STOP -> instruction = SLOPE_STOP
        }
        return createMsg(SLOPE_BODY, instruction, 0)
    }

    fun wheels(verticalSpeed: Double, horizontalSpeed: Double): String {
        return createMsg(WHEELS_BODY, 50 + round(50.0 * verticalSpeed).toInt(), 50 + round(50.0 * horizontalSpeed).toInt())
    }

    fun led(): String {
        return createMsg(LED_BODY, 0, 0)
    }

    fun gimbal(state: Int): String {
        return createMsg(GIMBAL_BODY, state, 0)
    }
}